### PR TITLE
REFAC(positional-audio): Port GTA5 to new Plugin API

### PR DESCRIPTION
**Pull request by vimpostor**
This updates the GTA 5 positional audio plugin to work with the latest Steam version and also ports the whole plugin to the new Plugin API.

### CODE REVIEW

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [inline plugins/gtav/gtav.cpp:80]:
```suggestion
	constexpr auto exeName = "GTA5.exe";
```

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [inline plugins/gtav/gtav.cpp:122]: Since the plugin is basically rewritten from scratch anyway, I would change this to:

```suggestion
	for (uint8_t i = 0; i < 3; ++i) {
```

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [inline plugins/gtav/gtav.cpp:14]: What about:

```suggestion
static procptr_t moduleBase = 0;
```

?

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [inline plugins/gtav/gtav.cpp:103]:
```suggestion
		const char *identity = "";
		const char *context  = "";
```

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [inline plugins/gtav/gtav.cpp:206]: This can be removed, we already initialize the arrays to 0 at the beginning of the function.

**Review by davidebeatrici — CHANGES_REQUESTED** [review]

**Comment by davidebeatrici** [comment]
Ideally initialization should be moved to a class, as I did for the `amongus` plugin:

https://github.com/mumble-voip/mumble/blob/master/plugins/amongus/Game.cpp
https://github.com/mumble-voip/mumble/blob/master/plugins/amongus/Game.h

However, it's something we can do in future.

File: plugins/gtav/gtav.cpp
Comment (vimpostor) [reply plugins/gtav/gtav.cpp]: Done, note though that `identity` overshadows the global static variable `std::string identity` now, which isn't really pretty.

**Comment by vimpostor** [comment]
> Ideally initialization should be moved to a class, as I did for the `amongus` plugin:
> 
> https://github.com/mumble-voip/mumble/blob/master/plugins/amongus/Game.cpp
> https://github.com/mumble-voip/mumble/blob/master/plugins/amongus/Game.h
> 
> However, it's something we can do in future.

I have no problem doing this as part of this PR. I have to admit though, I kinda like plugins just being a simple single CPP file too.
But I also see the reasons against this. Right now there are three static global variables present, which could be handled as single state with a separate Game class.

**Comment by davidebeatrici** [comment]
Indeed. In addition to that, you can also write an initialization function that returns a specific error code.

For example, if the detected game version is not supported, the plugin should report that it's permanently non-functional (`MUMBLE_PDEC_ERROR_PERM`).

File: plugins/gtav/gtav.cpp
Comment (Krzmbrzl) [inline plugins/gtav/gtav.cpp:14]:
```suggestion
static procptr_t moduleBase = nullptr;
```

**Review by Krzmbrzl — COMMENTED** [review]
For my taste `auto` is a bit overused in this code. There are quite a few places where I can't deduce the type of a variable by looking at the code.
Imo `auto` should only be used for things like iterators or in case the type is explicitly part of the statement already (e.g. when using a cast or when explicitly invoking a constructor)

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [reply plugins/gtav/gtav.cpp:14]: It's not a pointer.

Comment (Krzmbrzl) [reply plugins/gtav/gtav.cpp:14]: Lol then the naming seems to be very misleading. `ptr` usually indicates pointer-type :eyes:

**Comment by vimpostor** [comment]
> Imo `auto` should only be used for things like iterators or in case the type is explicitly part of the statement already (e.g. when using a cast or when explicitly invoking a constructor)

What instances of auto usage do you mean specifically here? Imo all instances are either iterator stuff or the type name is self explanatory from the function on the right side of the assignment.

For example:
```c++
const auto &modules = proc->modules();
// -> modules is of type const Modules&
const auto &regions = iter->second.regions();
// -> regions is of type const MemoryRegions&
```

**Comment by Krzmbrzl** [comment]
> the type name is self explanatory from the function on the right side of the assignment.

This is the part that is bugging me. In order to figure out the type of the variable, I'll have to go and look up that function's definition instead of being to read off the type directly when reading the code as-is.
I mean of course there is no situation in which the type of a variable is not "self-explanatory" from the statement producing the type. However my point is that I don't want to have the reader having to parse the statement and go search for definitions just to figure out what the type of the produce3d value is.

This is because not everyone reading the code will have a full-blown IDE at hand that will be able to show the variable's type on hover and where that is definitely not available is here on GitHub which makes reviewing PRs very cumbersome.

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [reply plugins/gtav/gtav.cpp:14]: In this case `ptr` means that the variable can be used to store a memory address.

**Comment by vimpostor** [comment]
> In order to figure out the type of the variable, I'll have to go and look up that function's definition

That's not what I meant. I would argue that when I see the method `->modules()`, it is almost as obvious that this returns type `Modules` as a `->toInt()` method returns an int. There is no need at all to look up the function's definition for this.

Anyways, I did what you said. I hope the rest of the `auto` usage is okay, otherwise tell me what instances you want to have replaced specifically.


@davidebeatrici I have completed the refactorization to a separate `Game` class. Now there is only a single global static variable.
I have left the new refactorization as a new commit, but I can squash them if you prefer.

Another thing I changed that I don't want to hide (as it is squashed into the last commit), is that there are no longer duplicate positional audio data vectors created. Instead everything is done in-place and the transformation between game coordinates and Mumble is done with a (imho) more elegant `std::swap` now. Loop unrolling of the compiler should ensure that there is no performance penalty.

File: plugins/gtav/Game.h
Comment (davidebeatrici) [inline plugins/gtav/Game.h:14]:
```suggestion
	Game(const procid_t id, const std::string &name);
```

Or:

```suggestion
	Game(const uint64_t_t id, const std::string &name);
```

File: plugins/gtav/gtav.cpp
Comment (davidebeatrici) [inline plugins/gtav/gtav.cpp:116]:
```suggestion
		// Swap Y with Z component
```

**Review by davidebeatrici — CHANGES_REQUESTED** [review]

**Comment by davidebeatrici** [comment]
The [`std::swap`](https://en.cppreference.com/w/cpp/algorithm/swap) solution is brilliant, I wasn't aware of that!

File: plugins/gtav/Game.h
Comment (vimpostor) [reply plugins/gtav/Game.h]: Settled for the first option. BTW the equivalent constructor of the AmongUS Game class does also not have a std::string reference. But doesn't really matter that much due to SSO.

File: plugins/gtav/Game.cpp
Comment (davidebeatrici) [inline plugins/gtav/Game.cpp:1]:
```suggestion
// Copyright 2021 The Mumble Developers. All rights reserved.
```

Comment (davidebeatrici) [inline plugins/gtav/Game.cpp:1]:
```suggestion
// Copyright 2021 The Mumble Developers. All rights reserved.
```

File: plugins/gtav/Game.cpp
Comment (davidebeatrici) [inline plugins/gtav/Game.cpp:91]: I would move this block in `Game::init()` (and move the variables into the class itself).

File: plugins/gtav/Game.cpp
Comment (davidebeatrici) [inline plugins/gtav/Game.cpp:37]: This block can be removed if you implement the suggestion below.

Comment (vimpostor) [reply plugins/gtav/Game.cpp]: I quite like that!

**Comment by vimpostor** [comment]
Like before, to ease iterative review I have added another separate commit.
Before merging, I can squash them together.

**Review by davidebeatrici — APPROVED** [review]
Perfect!

File: plugins/gtav/Game.h
Comment (Krzmbrzl) [inline plugins/gtav/Game.h:7]:
```suggestion
#ifndef GTAV_GAME_H_
#define GTAV_GAME_H_
```
(Just a convention we follow for the Mumble source code)

**Review by Krzmbrzl — APPROVED** [review]

**Comment by Krzmbrzl** [comment]
> That's not what I meant. I would argue that when I see the method ->modules(), it is almost as obvious that this returns type Modules as a ->toInt() method returns an int. There is no need at all to look up the function's definition for this.

I would argue that the situation is a bit different as `int` is a standard datatype that everyone knows about whereas `Modules` is not. Therefore there is still some ambiguity. For instance `->modules()` could just as well return something like `std::vector<Module>` or any other kind of list.
Only when you know that there exists a type `Modules`, will you be able to have a clue that this might be what the function is returning.

Anyway, thanks for changing these parts of the code. The rest LGTM :+1:

File: plugins/gtav/Game.h
Comment (vimpostor) [reply plugins/gtav/Game.h:7]: The `_H_` suffix was missing in amongus/Game.h too.

**Comment by vimpostor** [comment]
Do not merge yet. After moving the address initialization code to the plugin init code, there is a new bug where the plugin would rapidly link and unlink when running GTA, but not being online (because init returns ok, but then fetchpositionaldata returns error, because the character is not ingame right now).

File: plugins/gtav/Game.cpp
Comment (vimpostor) [reply plugins/gtav/Game.cpp]: > This block can be removed if you implement the suggestion below ([#5155 (comment)](https://github.com/mumble-voip/mumble/pull/5155#discussion_r661852945)).
Unfortunately I think there is no other way than to fetch positional data already in the init method, to fix the rapid linking and unlinking when the person is not ingame currently (see https://github.com/mumble-voip/mumble/pull/5155#issuecomment-872094014)

What do you say? I think we will have to add back that part, but can still keep address init in the init method.

**Comment by vimpostor** [comment]
I have now readded the initial check for meaningful data to fix the issue.
```diff
diff --git a/plugins/gtav/Game.cpp b/plugins/gtav/Game.cpp
index 9b8fd0d6b..75f4da3ea 100644
--- a/plugins/gtav/Game.cpp
+++ b/plugins/gtav/Game.cpp
@@ -67,7 +67,12 @@ Mumble_PositionalDataErrorCode Game::init() {
                return MUMBLE_PDEC_ERROR_PERM;
        }
 
-       return MUMBLE_PDEC_OK;
+       // Check if we can get meaningful data from it
+       float apos[3], afront[3], atop[3], cpos[3], cfront[3], ctop[3];
+       const char *identity = "";
+       const char *context  = "";
+
+       return fetchPositionalData(apos, afront, atop, cpos, cfront, ctop, &context, &identity);
 }
 
 Mumble_PositionalDataErrorCode Game::fetchPositionalData(float *avatarPos, float *avatarDir, float *avatarAxis,
```

I think there is no way around this.

File: plugins/gtav/Game.cpp
Comment (Krzmbrzl) [reply plugins/gtav/Game.cpp:75]: Another solution could be to pretend to still be able to deliver positional data even if the player is not in-game yet (always sending (0,0,0) as the position) thus preventing the unlink :thinking:

Comment (vimpostor) [reply plugins/gtav/Game.cpp:75]: > Another solution could be to pretend to still be able to deliver positional data even if the player is not in-game yet (always sending (0,0,0) as the position) thus preventing the unlink

No that wouldn't really work. People would still hear you positonally, but your voice coming from a very weird point of space. (0,0,0) does not disable positional audio, it is just an arbitrary fixed point of space and the only case where that would not behave weirdly is if all Mumble users have (0,0,0) set. In any other case that would not work.


And after all, probing positional data once in init is not such a big of a deal. It would have been a nice extra if that part could have been removed, but that didn't work.

**Comment by davidebeatrici** [comment]
You could write a separate function that returns a `bool` (`true` for success and `false` for failure) and takes as reference a structure containing the data read from the game.

**Comment by vimpostor** [comment]
> You could write a separate function that returns a `bool` (`true` for success and `false` for failure) and takes as reference a structure containing the data read from the game.

What would be the benefit of this? One would still need to probe for positional data once at the beginning, so I don't understand what one would get from this. The call for `fetchPositionalData` already returns some kind of `bool`, except that it is the direct status code.

**Comment by davidebeatrici** [comment]
In this case it wouldn't make much sense, as we set Mumble's vectors directly (the plugin is not aware of the game's structures right now).

**Comment by vimpostor** [comment]
Will this be available for 1.4 users? If I remember correctly, plugins can now be updated independently from the main release right?

**Comment by davidebeatrici** [comment]
Correct, this can be backported.

**Comment by Krzmbrzl** [comment]
Backport is available at #5162 

Thanks for updating the plugin :+1:

**Comment by davidebeatrici** [comment]
Thank you very much for your contribution!

**Pull request merged**
